script/copy_deduplicate

```diff
@@ -266,49 +270,85 @@ def copy_join_parts(client, stable_table, query_jobs):
+    if only_tables and not contains_glob(only_tables):
+        # skip list calls when only_tables exists and contains no globs
+        return [f"{project_id}.{t}" for t in only_tables]
+    if only_tables and not contains_glob(glob_dataset(t) for t in only_tables):
+        # skip list_datasets call when only_tables exists and datasets contain no globs
+        live_datasets = {f"{project_id}.{glob_dataset(t)}" for t in only_tables}
+    else:
+        live_datasets = [
+            d.reference
+            for d in client.list_datasets(project_id)
+            if d.dataset_id.endswith("_live")
+        ]
+    live_tables = [
+        f"{t.dataset_id}.{t.table_id}"
+        for tables in pool.map(client.list_tables, live_datasets)
+        for t in tables
+    ]
+    if only_tables:
+        match = compile_glob_patterns(only_tables).match
+        live_tables = [t for t in live_tables if glob_predicate(match, t, "only")]
+    if except_tables:
+        match = compile_glob_patterns(except_tables).match
+        live_tables = [t for t in live_tables if not glob_predicate(match, t, "except")]
+    return [f"{project_id}.{t}" for t in live_tables]
+
+
 def main():
     args = parser.parse_args()
 
     client = bigquery.Client()
-    live_datasets = [
-        d
-        for d in client.list_datasets(args.project_id)
-        if d.dataset_id.endswith("_live")
-    ]
 
-    job_args = []
-
-    for live_dataset in live_datasets:
-        stable_dataset_id = live_dataset.dataset_id[:-5] + "_stable"
-        stable_dataset = client.dataset(stable_dataset_id, args.project_id)
-        for live_table in client.list_tables(live_dataset.reference):
-            live_table_id = live_table.table_id
-            live_table_spec = f"{live_table.dataset_id}.{live_table_id}"
-            stable_table = stable_dataset.table(f"{live_table_id}${args.date:%Y%m%d}")
-            if args.except_tables is not None and any(
-                fnmatch(live_table_spec, pattern) for pattern in args.except_tables
-            ):
-                print(f"Skipping {live_table_spec} due to --except argument")
-                continue
-            if args.only_tables is not None and not any(
-                fnmatch(live_table_spec, pattern) for pattern in args.only_tables
-            ):
-                print(f"Skipping {live_table_spec} due to --only argument")
-                continue
-            sql = QUERY_TEMPLATE.format(live_table=sql_full_table_id(live_table))
-            job_args.extend(
-                (client, sql, stable_table, job_config)
-                for job_config in get_query_job_configs(
-                    client=client,
-                    stable_table=stable_table,
-                    date=args.date,
-                    dry_run=args.dry_run,
-                    slices=args.slices,
-                    priority=args.priority,
-                )
+    with ThreadPool(args.parallelism) as pool:
+        live_tables = list_live_tables(
```

Comment:
Do we risk running into throttling by parallelizing these bq api calls?

Reply:
the client automatically handles retries/backoff, and the relevant quotas are 100req/s per user, and 300 concurrent req/s per user, so unless parallelism is set very high this shouldn't be an issue.